SRM remove multiple ENSGs per peptide

jaclynbeck-sage opened the pull request:
This PR handles the fact that multiple Ensembl IDs map to the same UniProt ID, so SRM data had duplicates with Ensembl IDs that were mostly HASGs. 

I ended up doing a bunch of cleanup across multiple notebooks as a side-effect and added it to this PR. So this PR contains the following changes:
1. **SRM processing now removes HASGs that match to peptide UniProt IDs**. I've confirmed that the only Ensembl IDs remaining for peptides that had duplicates are the ones in bold in [the JIRA ticket](https://sagebionetworks.jira.com/browse/AG-1296), and no other genes were removed from the data set. 
2. I made a new file `preprocessing_utils.py` and pulled some duplicated code into functions in that file. Right now, that's requests to Biomart and filtering HASGs, but can be expanded on if necessary. It currently lives in the same folder as the pre-processing notebooks.
3. Changed SRM pre-processing, GWAS pre-processing, and gene metadata pre-processing to use these functions instead of duplicating code. 
4. Gene metadata processing code was pretty out of date and needed some updating to work again. 
5. **Gene metadata processing now removes HASGs that are pulled in by Biomart ONLY** (it doesn't touch the ones that come from our data source files since we haven't decided how to handle that yet). The resulting `gene_table_merged.feather` file and `gene_info.json` (post-transformation) will have FEWER genes than the last update.  
6. Updating the `gene_table_merged.feather` file **pulled in Ensembl release 111**. 
7. The version numbers in both config files were changed to use the newly-generated SRM and gene table files. 

I spot-checked the new `gene_info.json` file generated by CI and it looks good to me (see also [this JIRA comment](https://sagebionetworks.jira.com/browse/AG-1296?focusedCommentId=195692) for why some HASGs are still in `gene_info`). 

The diffs on the notebooks are awful to look at so I've linked the actual notebook files below:
igap [before](https://github.com/Sage-Bionetworks/agora-data-tools/blob/6039932ec4a503fa176bfb6e366a5a72d8cfce6b/data_analysis/agora/notebooks/preprocessing/AG-796_process_gwas.ipynb) and [after](https://github.com/Sage-Bionetworks/agora-data-tools/blob/1bf18f28b805bdd889e0188e60a72600ebafaea3/data_analysis/agora/notebooks/preprocessing/AG-796_process_gwas.ipynb)
srm [before](https://github.com/Sage-Bionetworks/agora-data-tools/blob/620c094d4e73eb37f9797f329e8009173d401269/data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb) and [after](https://github.com/Sage-Bionetworks/agora-data-tools/blob/1bf18f28b805bdd889e0188e60a72600ebafaea3/data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb)
gene annotations [before](https://github.com/Sage-Bionetworks/agora-data-tools/blob/62fda2f739aec4c6aa1ac4c2ff638e3fa505cc39/data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb) and [after](https://github.com/Sage-Bionetworks/agora-data-tools/blob/1bf18f28b805bdd889e0188e60a72600ebafaea3/data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb)

jaclynbeck-sage committed:
Updated SRM diffexp pre-processing to remove HASGs. Created preprocessing util functions to do this for SRM and IGAP, and updated IGAP pre-processing to use these functions

jaclynbeck-sage committed:
Bumped up versions of SRM data and gene_metadata in the config files. Updated the gene_metadata preprocessing script to work with recent changes to ADT

jaclynbeck-sage committed:
Fixed some SonarCloud issues

jaclynbeck-sage committed:
Fixed a few more SonarCloud issues

jaclynbeck-sage committed:
Fixed regex for filtering HASGs

jaclynbeck-sage committed:
Fixed bug in gene annotations pre-processing where genes from biodomains weren't being added

jaclynbeck-sage committed:
Bumped version of gene_metadata in config to use new file

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-796_process_gwas.ipynb:
```diff
@@ -317,25 +317,20 @@
-    "response = requests.get(url = 'http://www.ensembl.org/biomart/martservice', params = {'query': query})\n",
+    "result = preprocessing_utils.manual_query_biomart(\n",
+    "    attributes=attributes, filters=filters\n",
+    ")\n",
```
Here is where I moved biomart query code to preprocessing_utils.py

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-796_process_gwas.ipynb:
```diff
@@ -372,15 +358,10 @@
-    "    # Keep rows if they have a numerical chromosome name, or have X, Y, or MT\n",
-    "    lambda row: re.match(regex, row) is not None if type(row) == str else True # Always true for numbers \n",
+    "result = preprocessing_utils.filter_hasgs(\n",
+    "    df=result, chromosome_name_column=\"chromosome_name\"\n",
```
Here is where I moved the HASG filter code to preprocessing_utils.py

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb:
```diff
@@ -41,7 +41,8 @@
     "import synapseclient\n",
-    "import agoradatatools.etl.extract as extract"
+    "import agoradatatools.etl.extract as extract\n",
+    "import preprocessing_utils"
```
Most of the changes in this file are from the black formatter. I'll highlight actual code changes with comments.

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb:
```diff
@@ -442,8 +470,393 @@
+   "id": "eee4be20",
+   "metadata": {},
+   "source": [
+    "Filter out HASGs by querying Biomart for chromosome information and removing genes with HASG chromosomes notation. "
```
NEW code starts here to filter HASGs

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb:
```diff
@@ -442,8 +470,393 @@
+   ],
+   "source": [
+    "ensembl_ids = ensembl_ids.loc[ensembl_ids[\"ENSG\"].isin(valid_ids[\"ENSG\"])]\n",
+    "ensembl_ids"
```
End new code

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:
```diff
@@ -101,38 +103,97 @@
    "source": [
     "# Sometimes Biomart doesn't respond and the command needs to be sent again. Try up to 5 times.\n",
     "for T in range(5):\n",
     "    try:\n",
```
This chunk is a code update

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:
```diff
@@ -101,38 +103,97 @@
+    "    df=ensembl_ids_df, chromosome_name_column=\"chromosome_name\"\n",
+    ")\n",
+    "print(str(ensembl_ids_df.shape[0]) + \" genes remaining after HASG filtering.\")"
    ]
```
End code update

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:
```diff
@@ -241,73 +317,86 @@
+    "            print(file + \" has an NaN Ensembl ID\")\n",
+    "    else:\n",
+    "        print(\"WARNING: no Ensembl ID column found for \" + file + \"!\")"
    ]
```
The couple changes above are a code update

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:
```diff
@@ -241,73 +317,86 @@
+    "ensembl_ids_df = pd.concat([ensembl_ids_df, missing_df])\n",
+    "ensembl_ids_df = ensembl_ids_df.dropna(subset=[\"ensembl_gene_id\"])\n",
+    "print(len(ensembl_ids_df))"
    ]
```
The couple lines above are a code update

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:
```diff
@@ -710,110 +783,117 @@
-    "gene_table_merged = transform.standardize_values(gene_table_merged)\n",
+    "gene_table_merged = utils.standardize_column_names(gene_table_merged)\n",
+    "gene_table_merged = utils.standardize_values(gene_table_merged)\n",
     "\n",
```
Code update above

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:
```diff
@@ -831,19 +911,37 @@
+    "gene_table_merged[\"alias\"] = gene_table_merged[\"alias\"].apply(\n",
+    "    lambda row: list(set(flatten(row)))\n",
+    ")"
    ]
```
Code update above

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:
```diff
@@ -1782,235 +1409,88 @@
      "output_type": "execute_result"
     }
    ],
    "source": [
```
Begin code update chunk

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:
```diff
@@ -1782,235 +1409,88 @@
+    ")\n",
+    "gene_table_merged = keep_df.reset_index(drop=True)\n",
+    "gene_table_merged.tail(n=10)"
    ]
```
End code update chunk

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:
```diff
@@ -3091,21 +2594,282 @@
+    "]\n",
+    "\n",
+    "gene_table_merged"
    ]
```
Code update above

jaclynbeck-sage committed:
Removed a duplicated chunk of code from gene annotation preprocessing

jaclynbeck-sage commented on data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:
```diff
@@ -3119,12 +2874,16 @@
+    ")\n",
+    "gene_table_merged\n",
+    "gene_table_merged.to_feather(\"../../output/gene_table_merged_GRCh38.p14.feather\")"
    ]
```
Code update above

JessterB commented on data_analysis/agora/notebooks/preprocessing/preprocessing_utils.py:
```diff
@@ -0,0 +1,79 @@
+    """Filters human alternative sequence genes (HASGs) from a data frame by using a regex to
+    identify them for removal. Valid genes will either have a numerical chromosome name or have
+    "X", "Y", or "MT" as the chromosome name. All other chromosome name formats correspond to
+    HASGs, pathches, and other genes that should be removed from the list.
```
nitpick - 'pathches' looks like a typo

jaclynbeck-sage replied:
Yeah that's definitely a typo... oops. Will fix.

JessterB commented on data_analysis/agora/notebooks/preprocessing/preprocessing_utils.py:
```diff
@@ -0,0 +1,79 @@
+        df (pd.DataFrame): a data frame as returned from a Biomart query, that must have a column
+                           containing chromosome names. Biomart already returns chromosome names
+                           that fit the above regex pattern, but any manual creation of chromosome
+                           names should be formatted properly so the regex works before using this
```
curiosity - who is manually creating chromosome names, and why are they doing that?

jaclynbeck-sage replied:
Nobody is doing this. I don't know why, but I guess I added this to make the function sound 'generic'... Do you want me to just remove that part of the comment?

JessterB reviewed (approved):
This lgtm, although I did not exhaustively review the changes to the notebooks...nor am I at all sure that I'm competent to do so. I'm wondering if there's anyone else on our team that might be capable and willing to review your more scientific/specialized code...I know Laura is interested in developing her python skills, so maybe we could convince her that it would be a good learning experience!

jaclynbeck-sage committed:
Merge branch 'dev' of github.com:Sage-Bionetworks/agora-data-tools into jbeck/AG-1296/srm_remove_multi_ensgs

jaclynbeck-sage committed:
Fixed typo, removed confusing comment

jaclynbeck-sage commented:
I fixed the typo and also removed the comment about manually formatting chromosome names to avoid confusion. @JessterB do you want me to ask Laura to review the analysis part of the code before merging this PR?

JessterB commented:
@jaclynbeck-sage Having Laura review the analysis would be great if she is willing and has the time. I don't want to block merging this for more than ~1 week though, since I'd like to get the updated SRM data ready for release.

jaclynbeck-sage commented:
@JessterB I asked Laura and she said she's not at all familiar with Python. So I just described to her the steps I'm taking to process the data and she said it seemed sound. Not the same as a code review but, hopefully good enough for now?

JessterB commented:
@jaclynbeck-sage Yes, that works. Thanks!
